### PR TITLE
Add more type hints

### PR DESCRIPTION
**Pull request by alexkahan**
Add type hints to multiple design pattern examples

Related to https://github.com/faif/python-patterns/issues/373

### CODE REVIEW

File: patterns/behavioral/publish_subscribe.py [hunk 1]
```diff
@@ -5,47 +5,50 @@
 """
 
 
+from __future__ import annotations
+
+
 class Provider:
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/behavioral/publish_subscribe.py:12]: pointless

File: patterns/behavioral/publish_subscribe.py [hunk 2]
```diff
@@ -5,47 +5,50 @@
...
+    def update(self) -> None:
         for msg in self.msg_queue:
             for sub in self.subscribers.get(msg, []):
                 sub.run(msg)
         self.msg_queue = []
 
 
 class Publisher:
-    def __init__(self, msg_center):
+    def __init__(self, msg_center: Provider) -> None:
```
Comment (JKearnsl) [inline patterns/behavioral/publish_subscribe.py:33]: pointless

File: patterns/behavioral/publish_subscribe.py [hunk 3]
```diff
@@ -5,47 +5,50 @@
...
         self.provider = msg_center
 
-    def publish(self, msg):
+    def publish(self, msg: str) -> None:
         self.provider.notify(msg)
 
 
 class Subscriber:
-    def __init__(self, name, msg_center):
+    def __init__(self, name: str, msg_center: Provider) -> None:
```
Comment (JKearnsl) [inline patterns/behavioral/publish_subscribe.py:41]: pointless

File: patterns/behavioral/state.py [hunk 4]
```diff
@@ -22,43 +23,42 @@ def scan(self):
 
 
 class AmState(State):
-    def __init__(self, radio):
+    def __init__(self, radio: Radio) -> None:
```
Comment (JKearnsl) [inline patterns/behavioral/state.py:26]: pointless

File: patterns/behavioral/state.py [hunk 5]
```diff
@@ -22,43 +23,42 @@ def scan(self):
...
 
-    def toggle_amfm(self):
+    def toggle_amfm(self) -> None:
         print("Switching to FM")
         self.radio.state = self.radio.fmstate
 
 
 class FmState(State):
-    def __init__(self, radio):
+    def __init__(self, radio: Radio) -> None:
```
Comment (JKearnsl) [inline patterns/behavioral/state.py:38]: pointless

File: patterns/behavioral/state.py [hunk 6]
```diff
@@ -22,43 +23,42 @@ def scan(self):
...
         print("Switching to AM")
         self.radio.state = self.radio.amstate
 
 
 class Radio:
-
     """A radio.     It has a scan button, and an AM/FM toggle switch."""
 
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/behavioral/state.py:52]: pointless

File: patterns/creational/borg.py [hunk 7]
```diff
@@ -38,12 +38,12 @@
 class Borg:
     _shared_state: Dict[str, str] = {}
 
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/creational/borg.py:41]: pointless

File: patterns/creational/borg.py [hunk 8]
```diff
@@ -38,12 +38,12 @@
...
         self.__dict__ = self._shared_state
 
 
 class YourBorg(Borg):
-    def __init__(self, state=None):
+    def __init__(self, state: str = None) -> None:
```
Comment (JKearnsl) [inline patterns/creational/borg.py:46]: pointless

File: patterns/creational/borg.py [hunk 9]
```diff
@@ -52,7 +52,7 @@ def __init__(self, state=None):
             if not hasattr(self, "state"):
                 self.state = "Init"
 
-    def __str__(self):
+    def __str__(self) -> str:
```
Comment (JKearnsl) [inline patterns/creational/borg.py:55]: pointless

File: patterns/creational/builder.py [hunk 10]
```diff
@@ -34,7 +34,7 @@ class for a building, where the initializer (__init__ method) specifies the
 
 # Abstract Building
 class Building:
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/creational/builder.py:37]: pointless

File: patterns/creational/builder.py [hunk 11]
```diff
@@ -44,24 +44,24 @@ def build_floor(self):
     def build_size(self):
         raise NotImplementedError
 
-    def __repr__(self):
+    def __repr__(self) -> str:
```
Comment (JKearnsl) [inline patterns/creational/builder.py:47]: pointless

File: patterns/creational/builder.py [hunk 12]
```diff
@@ -72,19 +72,19 @@ def build_size(self):
 
 
 class ComplexBuilding:
-    def __repr__(self):
+    def __repr__(self) -> str:
```
Comment (JKearnsl) [inline patterns/creational/builder.py:75]: pointless

File: patterns/fundamental/delegation_pattern.py [hunk 13]
```diff
@@ -28,7 +28,7 @@ class Delegator:
     AttributeError: 'Delegate' object has no attribute 'do_anything'
     """
 
-    def __init__(self, delegate: Delegate):
+    def __init__(self, delegate: Delegate) -> None:
```
Comment (JKearnsl) [inline patterns/fundamental/delegation_pattern.py:31]: pointless

File: patterns/fundamental/delegation_pattern.py [hunk 14]
```diff
@@ -44,7 +44,7 @@ def wrapper(*args, **kwargs):
 
 
 class Delegate:
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/fundamental/delegation_pattern.py:47]: pointless

File: patterns/other/blackboard.py [hunk 15]
```diff
@@ -8,30 +8,33 @@
...
 import abc
 import random
 
+from typing import List
+
 
 class Blackboard:
-    def __init__(self):
-        self.experts = []
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/other/blackboard.py:20]: pointless "-> None" in __init__()

File: patterns/other/blackboard.py [hunk 16]
```diff
@@ -8,30 +8,33 @@
...
         }
 
-    def add_expert(self, expert):
+    def add_expert(self, expert: AbstractExpert) -> None:
         self.experts.append(expert)
 
 
 class Controller:
-    def __init__(self, blackboard):
+    def __init__(self, blackboard: Blackboard) -> None:
```
Comment (JKearnsl) [inline patterns/other/blackboard.py:34]: pointless "-> None" in init()

File: patterns/other/blackboard.py [hunk 17]
```diff
@@ -44,7 +47,7 @@ def run_loop(self):
 
 
 class AbstractExpert(metaclass=abc.ABCMeta):
-    def __init__(self, blackboard):
+    def __init__(self, blackboard: Blackboard) -> None:
```
Comment (JKearnsl) [inline patterns/other/blackboard.py:50]: pointless "-> None" in init()

File: patterns/structural/decorator.py [hunk 18]
```diff
@@ -28,30 +28,30 @@
 class TextTag:
     """Represents a base text tag"""
 
-    def __init__(self, text):
+    def __init__(self, text: str) -> None:
```
Comment (JKearnsl) [inline patterns/structural/decorator.py:31]: pointless "-> None" in init()

File: patterns/structural/decorator.py [hunk 19]
```diff
@@ -28,30 +28,30 @@
...
-    def render(self):
+    def render(self) -> str:
         return self._text
 
 
 class BoldWrapper(TextTag):
     """Wraps a tag in <b>"""
 
-    def __init__(self, wrapped):
+    def __init__(self, wrapped: TextTag) -> None:
```
Comment (JKearnsl) [inline patterns/structural/decorator.py:41]: pointless "-> None" in init()

File: patterns/structural/decorator.py [hunk 20]
```diff
@@ -28,30 +28,30 @@
...
-    def render(self):
+    def render(self) -> str:
         return f"<b>{self._wrapped.render()}</b>"
 
 
 class ItalicWrapper(TextTag):
     """Wraps a tag in <i>"""
 
-    def __init__(self, wrapped):
+    def __init__(self, wrapped: TextTag) -> None:
```
Comment (JKearnsl) [inline patterns/structural/decorator.py:51]: pointless "-> None" in init()

File: patterns/structural/front_controller.py [hunk 21]
```diff
@@ -5,23 +5,27 @@
...
 
 class TabletView:
-    def show_index_page(self):
+    def show_index_page(self) -> None:
         print("Displaying tablet index page")
 
 
 class Dispatcher:
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/structural/front_controller.py:24]: pointless "-> None" in init()

File: patterns/structural/front_controller.py [hunk 22]
```diff
@@ -39,10 +43,10 @@ def dispatch(self, request):
 class RequestController:
     """front controller"""
 
-    def __init__(self):
+    def __init__(self) -> None:
```
Comment (JKearnsl) [inline patterns/structural/front_controller.py:46]: pointless "-> None" in init()

**Review by JKearnsl — CHANGES_REQUESTED** [review]
Using the "-> None" hint in __init__ is meaningless, since __init__ always returns None and we can't use that anywhere.

The same goes for "-> str" in __repr__ and __str__ as we already deliberately know that these methods always return a string -> this is pointless

**Comment by faif** [comment]
We have already discussed about the `None` hints in another issue, and agreed to keep them, since this seems to be the convention. I'm also not a big fan of them, but I won't go against the syntax conventions of type hints.

**Comment by JKearnsl** [comment]
> We have already discussed about the `None` hints in another issue, and agreed to keep them, since this seems to be the convention. I'm also not a big fan of them, but I won't go against the syntax conventions of type hints.

OK. 
I'm just going from the logical meaning of the clues, not from mindless beliefs.

**Comment by faif** [comment]
Hi @alexkahan, can you look into the linter issues? Thanks.

**Comment by alexkahan** [comment]
> Hi @alexkahan, can you look into the linter issues? Thanks.

Sorry for the delay. Should be fixed, seemed like it was having an issue with the List type hint

**Review by JKearnsl — APPROVED** [review]

**Pull request merged**